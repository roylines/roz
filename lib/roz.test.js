const sleep = require('await-sleep');
const {Lambda} = require('aws-sdk');
const {handler} = require('./roz');
const {open, get} = require('./bastion');
const {send} = require('./message');
const {warn} = require('lambda-log');

jest.mock('./bastion');
jest.mock('./message');
jest.mock('await-sleep');
jest.mock('await-sleep');
jest.mock('aws-sdk');
jest.mock('lambda-log');
jest.mock('messaging-api-telegram');

describe('handler', () => {
  const context = {
    invokedFunctionArn: 'ARN',
  };
  describe('from api gateway', () => {
    it('should return 200 for unknown user', async () => {
      warn.mockReset();
      process.env.TELEGRAM_USER = 88;
      const body = {message: {from: {id: 42}}};
      const response = await handler({body: JSON.stringify(body)}, context);

      expect(response).toMatchSnapshot({statusCode: 200});
      expect(warn.mock.calls[0][0]).toEqual('unauthorised user');
    });

    it('should reinvoke for known user', async () => {
      process.env.TELEGRAM_USER = 42;
      const promise = jest.fn().mockResolvedValue({});
      const invoke = jest.fn().mockReturnValue({promise});
      Lambda.prototype.invoke = invoke;
      const body = {message: {text: 'hello', from: {id: 42}}};

      const response = await handler({body: JSON.stringify(body)}, context);

      expect(response).toMatchSnapshot({statusCode: 200});
      expect(invoke.mock.calls).toMatchSnapshot();
    });

    it('should return 500 for throws', async () => {
      const response = await handler({});
      expect(response).toMatchSnapshot({statusCode: 500});
    });
  });
  describe('from reinvoke', () => {
    it('should deal with hello', async () => {
      send.mockReset();
      const evt = {text: 'hello'};
      const response = await handler(evt, context);
      expect(response).toMatchSnapshot();
      expect(send.mock.calls).toMatchSnapshot();
    });
    it('should deal with break glass', async () => {
      send.mockReset();
      const evt = {text: 'break glass'};
      const response = await handler(evt, context);
      expect(response).toMatchSnapshot();
      expect(send.mock.calls).toMatchSnapshot();
      expect(open.mock.calls).toMatchSnapshot();
      expect(get.mock.calls).toMatchSnapshot();
    });
  });
});
