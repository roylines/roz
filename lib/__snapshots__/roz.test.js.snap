// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler from api gateway should reinvoke for known user 1`] = `
Object {
  "statusCode": 200,
}
`;

exports[`handler from api gateway should reinvoke for known user 2`] = `
Array [
  Array [
    Object {
      "FunctionName": "ARN",
      "InvocationType": "Event",
      "Payload": "{\\"text\\":\\"hello\\",\\"from\\":{\\"id\\":42}}",
    },
  ],
]
`;

exports[`handler from api gateway should return 200 for unknown user 1`] = `
Object {
  "statusCode": 200,
}
`;

exports[`handler from api gateway should return 500 for throws 1`] = `
Object {
  "statusCode": 500,
}
`;

exports[`handler from reinvoke should deal with break glass 1`] = `undefined`;

exports[`handler from reinvoke should deal with break glass 2`] = `
Array [
  Array [
    Array [
      Array [
        "OK, I'm opening the bastion now.",
      ],
    ],
  ],
  Array [
    Array [
      Array [
        "I've created the server. Just waiting for it to create...",
      ],
    ],
  ],
  Array [
    Array [
      Array [
        "Right it's done. Don't do anything silly.",
        "I'm watching you...",
      ],
    ],
  ],
]
`;

exports[`handler from reinvoke should deal with break glass 3`] = `
Array [
  Array [],
]
`;

exports[`handler from reinvoke should deal with break glass 4`] = `Array []`;

exports[`handler from reinvoke should deal with hello 1`] = `undefined`;

exports[`handler from reinvoke should deal with hello 2`] = `
Array [
  Array [
    Array [
      Array [
        "I'm watching you...",
        "Always watching...",
        "Always!",
      ],
    ],
  ],
]
`;
